src/groundlight/internalapi.py

```diff
@@ -225,3 +235,125 @@ def _get_detector_by_name(self, name: str) -> Detector:
+        """Submits an image query to the API and returns the ID of the image query.
+        The image query will be associated to the inspection_id provided.
+        """
+        url = f"{self.configuration.host}/posichecks?inspection_id={inspection_id}&predictor_id={detector_id}"
+
+        headers = self._headers()
+        headers["Content-Type"] = "image/jpeg"
+
+        response = requests.request("POST", url, headers=headers, data=image.read())
+
+        if not is_ok(response.status_code):
+            logger.info(response)
+            raise InternalApiError(
+                status=response.status_code,
+                reason=f"Error submitting image query with inspection ID {inspection_id} on detector {detector_id}",
+                http_resp=response,
+            )
+
+        return response.json()["id"]
+
+    @RequestsRetryDecorator()
+    def start_inspection(self) -> str:
+        """Starts an inspection, returns the ID."""
+        url = f"{self.configuration.host}/inspections"
+
+        headers = self._headers()
+
+        response = requests.request("POST", url, headers=headers, json={})
+
+        if not is_ok(response.status_code):
+            raise InspectionError(f"Error starting inspection. Status code: {response.status_code}")
+
+        return response.json()["id"]
+
+    @RequestsRetryDecorator()
+    def update_inspection_metadata(self, inspection_id: str, user_provided_key, user_provided_value) -> None:
+        """Add/update inspection metadata with the user_provided_key and user_provided_value.
+
+        The API stores inspections metadata in two ways:
+           1) At the top level of the inspection with user_provided_id_key and user_provided_id_value. This is a
+              kind of "primary" piece of metadata for the inspection. Only one key/value pair is allowed at this level.
+           2) In the user_metadata field as a dictionary.  Multiple key/value pairs are allowed at this level.
+
+        The first piece of metadata presented to an inspection will be assumed to be the user_provided_id_key and
+        user_provided_id_value. All subsequent pieces metadata will be stored in the user_metadata field.
+
+        """
+        url = f"{self.configuration.host}/inspections/{inspection_id}"
+
+        headers = self._headers()
+
+        # Get inspection to see if user_provided_id_key is set
+        response = requests.request("GET", url, headers=headers)
+
+        if not is_ok(response.status_code):
+            raise InspectionError(
+                f"Error getting inspection details for inspection {inspection_id}. Status code: {response.status_code}"
+            )
+
+        payload = {}
+
+        # Set the user_provided_id_key and user_provided_id_value if they were not previously set.
+        response_json = response.json()
+        if not response_json["user_provided_id_key"]:
```

Comment:
I'm not 100% sure -- will the `"user_provided_id_key"` key always be in the response (even if the value is `None`)?

To be safe, we could do `if not response_json.get("user_provided_id_key")`, which won't fail if the key doesn't exist.

Reply:
It should always be there. I copied this line of code from roobik and it worked there. But just to be safe, I will use .get